haas/ext/auth/keystone.py

```diff
@@ -0,0 +1,64 @@
+"""Keystone authentication backend.
+
+This is a thin wrapper around the `keystonemiddleware` library.
+"""
+from keystonemiddleware.auth_token import filter_factory
+from flask import request
+from haas.flaskapp import app
+from haas.config import cfg
+from haas.model import Project
+from haas import auth, rest
+import logging
+import sys
+
+logger = rest.ContextLogger(logging.getLogger(__name__), {})
+
+
+class KeystoneAuthBackend(auth.AuthBackend):
+
+    def authenticate(self):
+        # keystonemiddleware makes auth info available from two places:
+        #
+        # 1. Variables in the wsgi environment
+        # 2. Extra HTTP headers.
+        #
+        # In general, the wsgi variable 'HTTP_FOO_BAR' is equivalent to the
+        # HTTP header 'Foo-Bar'.
+        #
+        # We use the wsgi environment's variables below; it shouldn't matter,
+        # but this way if something goes horribly wrong and arbitrary headers
+        # aren't stripped out, the client can't just inject these.
+        if request.environ['HTTP_X_IDENTITY_STATUS'] != 'Confirmed':
+            return False
+
+        project_id = request.environ['HTTP_X_PROJECT_ID']
+        if Project.query.filter_by(label=project_id).count() == 0:
+            logger.info("Successful authentication by Openstack project %r, "
+                        "but this project is not registered with HaaS",
+                        project_id)
+            return False
+
+        return True
+
+    def _have_project_access(self, project):
+        return project.label == request.environ['HTTP_X_PROJECT_ID']
+
+    def _have_admin(self):
+        return 'admin' in request.environ['HTTP_X_ROLES'].split(',')
```

Comment:
@henn With typical Openstack policy files, there is no such thing as a per-admin project.  At some point we could move to having our own policy file, in which case we can do whatever we want.  But for now, we should work with the default policy that most Openstack installations will use.  (And the cloud_admin stuff is for more complicated domain stuff, which we also probably shouldn't use by default)